tools/filetools.go

```diff
@@ -121,3 +126,250 @@ func VerifyFileHash(oid, path string) error {
+	// Ignore all git metadata including subrepos
+	excludePaths := []string{".git", filepath.Join("**", ".git")}
+	return FastWalkWithExcludeFiles(dir, ".gitignore", nil, excludePaths)
+}
+
+func fastWalkFromRoot(dir string, excludeFilename string,
+	includePaths, excludePaths []string, fiChan chan<- FastWalkInfo, errChan chan<- error) {
+
+	dirFi, err := os.Stat(dir)
+	if err != nil {
+		errChan <- err
+		return
+	}
+
+	// This waitgroup will be incremented for each nested goroutine
+	var waitg sync.WaitGroup
+
+	fastWalkItem(filepath.Dir(dir), dirFi, excludeFilename, includePaths, excludePaths, fiChan, errChan, &waitg)
+
+	waitg.Wait()
+	close(fiChan)
+	close(errChan)
+
+}
+
+// Main recursive implementation of fast walk
+// Increment waitg.Add(1) for each new goroutine launched internally
+func fastWalkItem(parentDir string, itemFi os.FileInfo, excludeFilename string,
+	includePaths, excludePaths []string, fiChan chan<- FastWalkInfo, errChan chan<- error,
+	waitg *sync.WaitGroup) {
+
+	fullPath := filepath.Join(parentDir, itemFi.Name())
+
+	if !FilenamePassesIncludeExcludeFilter(fullPath, includePaths, excludePaths) {
+		return
+	}
+
+	fiChan <- FastWalkInfo{ParentDir: parentDir, Info: itemFi}
+
+	if !itemFi.IsDir() {
+		// Nothing more to do if this is not a dir
+		return
+	}
+
+	if len(excludeFilename) > 0 {
+		possibleExcludeFile := filepath.Join(fullPath, excludeFilename)
+		if FileExists(possibleExcludeFile) {
+			var err error
+			excludePaths, err = loadExcludeFilename(possibleExcludeFile, fullPath, excludePaths)
+			if err != nil {
+				errChan <- err
+			}
+		}
+	}
+
+	// The absolute optimal way to scan would be File.Readdirnames but we
+	// still need the Stat() to know whether something is a dir, so use
+	// File.Readdir instead. Means we can provide os.FileInfo to callers like
+	// filepath.Walk as a bonus.
+	df, err := os.Open(fullPath)
+	if err != nil {
+		errChan <- err
+		return
+	}
+	defer df.Close()
+	jobSize := 100
+	for children, err := df.Readdir(jobSize); err == nil; children, err = df.Readdir(jobSize) {
+		// Parallelise all dirs, and chop large dirs into batches
+		waitg.Add(1)
+		go func(subitems []os.FileInfo) {
+			for _, childFi := range subitems {
+				fastWalkItem(fullPath, childFi, excludeFilename, includePaths, excludePaths, fiChan, errChan, waitg)
+			}
+			waitg.Done()
+		}(children)
+
+	}
+	if err != nil && err != io.EOF {
+		errChan <- err
+	}
+
+}
+
+// loadExcludeFilename reads the given file in gitignore format and returns a
+// revised array of exclude paths if there are any changes.
+// If any changes are made a copy of the array is taken so the original is not
+// modified
+func loadExcludeFilename(filename, parentDir string, excludePaths []string) ([]string, error) {
+
+	f, err := os.OpenFile(filename, os.O_RDONLY, 0644)
+	if err != nil {
+		return excludePaths, err
+	}
+	defer f.Close()
+
+	retPaths := excludePaths
+	modified := false
```

Comment:
I prefer the way it is, lol. If this is later changed to require a default value of `true`, then the diff of that update would just be:

```diff
-modified := false
+modified := true
```